### ISSUE TITLE
Redirect Return Type Error for GET Route Handler with 13.5.1

### ISSUE THREAD

**Issue by agrattan0820** [vercel/next.js#55623]
### Link to the code that reproduces this issue

https://github.com/agrattan0820/next-redirect-replication

### To Reproduce

1. npm install
2. npm run build
3. Should receive type error preventing working build

### Current vs. Expected behavior

As of now, any route handler receives a type error during build since it is looking for a `Response` being returned, however the redirect function returns `never`. The expected behavior should be that a GET route handler ending with a redirect does not receive a type error.

Code:
```typescript
import { redirect } from "next/navigation";

export async function GET(request: Request) {
  redirect("https://nextjs.org/");
}
```

Error:
```bash
src/app/api/route.ts
Type error: Route "src/app/api/route.ts" has an invalid export:
  "Promise<void>" is not a valid GET return type:
    Expected "Response | Promise<Response>", got "Promise<void>".
      Expected "Promise<Response>", got "Promise<void>".
        Expected "Response", got "void".
```

As shown in the docs, this interaction should be allowed (unless it was changed with 13.5 and I am unaware): https://nextjs.org/docs/app/building-your-application/routing/route-handlers#redirects

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:52 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.13.0
      npm: 9.6.5
      Yarn: 1.22.19
      pnpm: 8.6.11
    Relevant Packages:
      next: 13.5.1
      eslint-config-next: 13.5.1
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

<sub>[NEXT-1642](https://linear.app/vercel/issue/NEXT-1642/redirect-return-type-error-for-get-route-handler-with-1351)</sub>

**Comment by shalabi** [comment]
Is there a workaround?

**Comment by rodolphoasb** [comment]
Related issue #55604

**Comment by yasseralsaidi** [comment]
same here how we fix it?

**Comment by yasseralsaidi** [comment]
i fix mine by replace 
 return Error(error);
with
 return Response.json({ error: error.message }, { status: 500 });

**Comment by darthmaim** [comment]
You can add `Promise<never>` as return type:

```ts
export async function GET(request: Request): Promise<never> {
  redirect('https://nextjs.org/')
}
```

**Comment by djaydevs** [comment]
This issue also occurs with version 13.4.19

**Comment by agrattan0820** [comment]
> This issue also occurs with version 13.4.19

My project with 13.4.19 is able to build without a type error however, the issue only arose upon upgrading to 13.5 for me

**Comment by agrattan0820** [comment]
> You can add `Promise<never>` as return type:
> 
> ```ts
> export async function GET(request: Request): Promise<never> {
>   redirect('https://nextjs.org/')
> }
> ```

Thanks, this worked for me

**Comment by giovannetti-eric** [comment]
If I try the fix that infer this type `Promise<never>`, I get this TS error:

> A function returning 'never' cannot have a reachable end point.ts(2534)

So I'm quite out of luck...

**Comment by anthonyverducci** [comment]
Similar errors here after upgrading to `13.5.2`:

```
Type error: Route "src/app/api/cron/email-queue/route.ts" has an invalid export:
--
08:49:16.734 | "Promise<NextResponse<{ message: string; }> \| undefined>" is not a valid GET return type:
08:49:16.734 | Expected "Response \| Promise<Response>", got "Promise<NextResponse<{ message: string; }> \| undefined>".
08:49:16.734 | Expected "Promise<Response>", got "Promise<NextResponse<{ message: string; }> \| undefined>".
08:49:16.734 | Expected "Response", got "NextResponse<{ message: string; }> \| undefined".
08:49:16.734 | Expected "Response", got "undefined".
```
I haven't been able to find a workaround yet. Similar to @giovannetti-eric, `Promise<never>` throws a type error.

**Comment by tjhorwood** [comment]
I was getting the same exact issue on some of my api routes. It came down to the more nested if/then statements in the try/catch. Reducing the number of conditions in the catch seemed to resolve the problem for me.

For instance changing: 
```javascript
} catch (error) {
    const errorMessage = 'An error occured!';
    if (error instanceof Error) {
      const isDuplicate = error.message.includes(
        'Unique constraint failed on the fields: (`username`)',
      );
      if (isDuplicate) {
        return NextResponse.json(
          { message: 'Username is already taken, please try another one' },
          { status: 501 },
        );
      } else {
        return NextResponse.json({ message: errorMessage }, { status: 503 });
      }
    }
  }
``` 
to this: 
```javascript
catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : 'An error occured. Please check username and password.';
    return NextResponse.json(
      { message: errorMessage, ok: false },
      { status: 503 },
    );
  }
```

**Comment by huix11** [comment]
👍👍👍👍👍

**Comment by yaxhhere** [comment]
try using Promise<any> as a return type for time being.

**Comment by anthonyverducci** [comment]
Thanks @tjhorwood. My `try/catch` blocks don't have any nested `if/then` statements but I will keep that in mind!

@yaxhhere `Promise<any>` did the trick for now. Thank you!

**Comment by huozhi** [comment]
We had a fix in 13.5.3, can you try to upgrade to the latest version? Thanks!

**Comment by cvogl** [comment]
Unfortunately the patch was not a fix, at least not here.
tl;dr Code unchanged since July, breaking of build persists with Next 13.5.3 (but works with 13.4.19)

I have just upgraded to Next 13.5.3 and linter breaks build with the same error message as with 13.5.1:

Type error: Route "src/app/api/frontend/checksession/route.ts" has an invalid export:
"Promise<false | NextResponse<{ status: string; }>>" is not a valid GET return type:
Expected "void | Response | Promise<void | Response>", got "Promise<false | NextResponse<{ status: string; }>>".
Expected "Promise<void | Response>", got "Promise<false | NextResponse<{ status: string; }>>".
Expected "void | Response", got "false | NextResponse<{ status: string; }>".
Expected "void | Response", got "boolean".

Double-checked the build after downgrade to 13.4.19: still finishing successfully.
(Which is no surprise: the code in that module was not touched after August 2.)

For us, it's also unrelated to images.

Thank you for investigating!

**Comment by huozhi** [comment]
@cvogl `Promise<false | NextResponse<{ status: string; }>>` sounds like an unexpected return value, if your handler return `false` it might hang the response if code branches to that condition. Sounds like sth need to tweak a bit in your code. Can you file another new issue with reproduction?

**Comment by cvogl** [comment]
Sure, you're right. 

Using explicit annotation with Promise<any> as suggested by @yaxhhere worked around the issue for me, btw.
